### DISCUSSION TITLE
Ideas on Design Principles and Patterns

### DISCUSSION THREAD

**Discussion by SiruLee** [CSC207-2022F-UofT/course-project-zuck-s-children#77] [Ideas]
## Design Principles and Patterns
 Please freely share ideas and opinions on your own/others' codes regarding design principles and patterns to improve our project!

  ➡️ You can comment on pull requests or just throw suggestions on what design patterns would fit to certain functions!

**Comment by SanzSS** [comment]
@XiWang16, I think you can implement a makeshift builder for profile as you have numerous attributes that need to be set independently. The builder can set each of these attributes.

**Reply by XiWang16** [reply]
Good point. I have written code to implement a design pattern that accomplishes this task but I called it a factory. Now that you've suggested, I will do some research to review what each pattern does and make sure I'm naming and implementing my classes correctly.

**Comment by SanzSS** [comment]
@lancenigel, Consider implementing a strategy design pattern for your matchingalgorithm.

**Reply by lancenigel** [reply]
I've attempted a strategy design pattern but I don't think it quite works as I don't "use different variants of an algorithm within an object and  don't need to switch from one algorithm to another during runtime".

- referenced this article: https://www.linkedin.com/pulse/strategy-designpattern-majid-ahmadi#:~:text=Use%20the%20Strategy%20pattern%20when,way%20they%20execute%20some%20behavior.

**Comment by XiWang16** [comment]
@tankenji As Hafsa suggested, you can implement the Singleton design pattern for the UserDatabase because there should only be a single instance of the database to which all other classes should refer.

**Reply by tankenji** [reply]
@XiWang16 Yes, good idea Joy. It's much cleaner code to have UserDatabase only refer to one static instance so I will implement the Singleton design pattern. This will mean I will need to refactor some other parts of the code as well to make it compatible to this change including the main class. I will figure out how to work on implementing this with in the main method.

**Comment by nina-huangg** [comment]
Just a general design pattern to keep in mind for next time - observer pattern! This way any changes made by a subject entity can be observed by its observers and notified immediately. This could apply to the changes made for profile and the observer being the study spot class.
